### PR TITLE
Add tests for Float8Tensor at graph boundaries

### PR DESCRIPTION
**Pull request by drisspg**
# Summary
This is a copy with some tweaks of: https://github.com/pytorch-labs/float8_experimental/pull/166

### CODE REVIEW

File: test/test_compile.py [hunk 1]
```diff
@@ -76,5 +78,80 @@ def test_inductor(fullgraph, emulate: bool, linear_type: bool, dtype: torch.dtyp
...
+
+    @pytest.mark.xfail(reason="TODO: Fix this test, see TODO in MockLinear")
+    def test_float8_graph_output(self):
+        """Test that having Float8Tensor object as a graph output works"""
+        mod = self.MockLinear(graph_break=False).cuda()
+        compiled_mod = torch.compile(mod)
+        x = torch.randn(16, 16, device="cuda")
+        y_compiled = compiled_mod(x)
+
+        assert not isinstance(
```
Comment (drisspg) [inline test/test_compile.py:138]: lets call __unflatten__ and tree map

Comment (bdhirsh) [reply test/test_compile.py:138]: rip this assert haha. I promised Voz I would add this check more natively in dynamo here: https://github.com/pytorch/pytorch/issues/118211, so eventually you (hopefully) won't need to feel so paranoid

File: test/test_compile.py [hunk 2]
```diff
@@ -76,5 +78,80 @@ def test_inductor(fullgraph, emulate: bool, linear_type: bool, dtype: torch.dtyp
...
+            self.register_buffer("fp8_scale_x", torch.tensor(1.0))
+            self.graph_break = graph_break
+
+        def forward(self, x):
+            x_fp8 = Float8Tensor.to_float8(
+                x,
+                self.fp8_scale_x,
+                torch.float8_e4m3fn,
+                self.fp8_amax_x,
+                emulate=True,  # TODO: I set this to True so that people on A100 can test, but once fix is in, set to False
```
Comment (bdhirsh) [inline test/test_compile.py:95]: fwiw - I noticed that I was still getting weird errors on my A100 with this test even with `emulate` manually hardcoded to True in the tests. It seemed like inductor was still generating native triton fp8 code?
```
CompilationError: at 10:19:def triton_(in_ptr0, in_ptr1, out_ptr0, xnumel, XBLOCK : tl.constexpr):
    xnumel = 4
    xoffset = tl.program_id(0) * XBLOCK
    xindex = xoffset + tl.arange(0, XBLOCK)[:]
    xmask = xindex < xnumel
    x0 = xindex
    tmp0 = tl.load(in_ptr0 + (x0), xmask)
    tmp2 = tl.load(in_ptr1 + (0))
    tmp3 = tl.broadcast_to(tmp2, [XBLOCK])
    tmp1 = tmp0.to(tl.float32)
                   ^
AssertionError('fp8e4nv data type is not supported on CUDA arch < 89')
```

**Review by bdhirsh — APPROVED** [review]

File: test/test_compile.py (hunk 2 above)
Comment (drisspg) [reply test/test_compile.py:95]: that is weird, I wonder why triton doesn't support the type on a100 and just fail on math ops

File: test/test_compile.py [hunk 3]
```diff
@@ -76,5 +78,79 @@ def test_inductor(fullgraph, emulate: bool, linear_type: bool, dtype: torch.dtyp
     _test_compile_base("inductor", fullgraph, emulate, linear_type, dtype)
 
 
+class TestGraphBreaks:
```
Comment (voznesenskym) [inline test/test_compile.py:81]: (1) test_compile is a weird file when we have test/dynamo/*
(2) Can we use regular dynamo testing infra? compile counter, etc?
(3) 1 test class per test file please
(4) This class name is wrong? It's testing fp8 + dynamo, not testing graph breaks (I would imagine TestGraphBreaks would test things like deep dynamo internal working)

File: test/test_compile.py [hunk 4]
```diff
@@ -76,5 +78,79 @@ def test_inductor(fullgraph, emulate: bool, linear_type: bool, dtype: torch.dtyp
...
+        """Test that having Float8Tensor object as a graph input"""
+
+        def to_float(x):
+            return x.to_original_precision()
+
+        to_float = torch.compile(to_float)
+
+        mod = self.MockLinear(graph_break=False).cuda()
+        x = torch.randn(2, 2, device="cuda")
+        compiled_to_float = torch.compile(to_float)
```
Comment (voznesenskym) [inline test/test_compile.py:124]: to_float could hit a skip, and look exactly identical, and still return as compiled_to_float, and this test would pass.

**Review by voznesenskym — CHANGES_REQUESTED** [review]
Please use existing dynamo test infra.

File: test/test_compile.py (hunk 3 above)
Comment (drisspg) [reply test/test_compile.py:81]: 1.) We don't have test/dynamo/*
2.) Sure is there a pointer for this?
3.) I hadn't heard of this rule before is this in some style guide convention somewhere?
4.) I thought thats implied by being in Float8Experimetnal, no?

File: test/test_compile.py (hunk 4 above)
Comment (drisspg) [reply test/test_compile.py:124]: you are right thats a bug 120 should be removed

File: test/test_compile.py (hunk 3 above)
Comment (voznesenskym) [reply test/test_compile.py:81]: (1) I know, but you can put a test there that depends on this, and have CI install it, right? Alternatively, if you depend on dynamo, just pull the deps out

(2) Yes! Check any test under test/dynamo/*, you dont have to inherit from the base class, but it does cover things like .reset() for you nicely (which I think we missed here!) - grep for CompileCounter :)

(3) No, but I think it helps with organization. Its just a tiny nipick

(4) True, that is my mistake.

File: test/test_compile.py (hunk 4 above)
Comment (voznesenskym) [reply test/test_compile.py:124]: Sorry, thats not what I mean! I mean that if you dont use our test infra that counts frames, you could hit a skip, get a frame_count of 0, and this test would pass!

**Comment by drisspg** [comment]
@voznesenskym Updated to use CompileCounterWithBackend
